pkg/upgrade/upgrades/v24_3_sql_instances_add_draining_test.go

```diff
@@ -0,0 +1,200 @@
+		binaryVersion := roachpb.Version{Major: 28, Minor: 4}
+		const expiration = time.Minute
+
+		createInstance := func(id base.SQLInstanceID) {
+			session := makeSession()
+			session.StartTS = clock.Now()
+			session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
+			instance, err := storage.CreateNodeInstance(
+				ctx, session, rpcAddr, sqlAddr, locality, binaryVersion, roachpb.NodeID(id))
+			require.NoError(t, err)
+			require.Equal(t, id, instance.InstanceID)
+		}
+
+		const firstId = base.SQLInstanceID(1000)
+		createInstance(firstId)
+		upgrades.Upgrade(t, sqlDB, clusterversion.V24_3_SQLInstancesAddDraining, nil, false)
+		const secondId = base.SQLInstanceID(1001)
+		createInstance(secondId)
+
+		// Verify the rows using SQL
+		r := sqlutils.MakeSQLRunner(sqlDB)
+		r.CheckQueryResults(t,
+			`SELECT id, is_draining FROM system.sql_instances WHERE id >= 1000`,
+			[][]string{{"1000", "NULL"}, {"1001", "false"}})
+
+		// Verify the rows by decoding key-values
+		instances, err := storage.GetAllInstancesDataForTest(ctx)
+		require.NoError(t, err)
+		instancestorage.SortInstances(instances)
+		require.Equal(t, 3, len(instances)) // Instead with id 1 also exists
+		require.Equal(t, base.SQLInstanceID(1), instances[0].GetInstanceID())
+		require.Equal(t, firstId, instances[1].GetInstanceID())
+		require.Equal(t, false, instances[1].IsDraining)
+		require.Equal(t, secondId, instances[2].GetInstanceID())
+		require.Equal(t, false, instances[2].IsDraining)
+	})
+
+	t.Run("old-readers-reading-new-schema-values", func(t *testing.T) {
+		ts, sqlDB, kvDB := setupServer(t)
+		defer ts.Stopper().Stop(ctx)
+		stopper, storage, clock := setup(t, ts, kvDB)
+		defer stopper.Stop(ctx)
+		const rpcAddr = "rpcAddr"
+		const sqlAddr = "sqlAddr"
+		locality := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "test"}, {Key: "az", Value: "a"}}}
+		binaryVersion := roachpb.Version{Major: 28, Minor: 4}
+		const expiration = time.Minute
+		// createInstance bypasses version guard and directly writes the key-value
+		// to storage as it uses CreateInstanceDataForTest instead of
+		// CreateNodeInstance.
+		createInstance := func(id base.SQLInstanceID, encodeIsDraining bool, isDraining bool) {
+			session := makeSession()
+			err := storage.CreateInstanceDataForTest(
+				ctx, enum.One, id, rpcAddr, sqlAddr,
+				session.SessionID, clock.Now().Add(expiration.Nanoseconds(), 0),
+				locality, binaryVersion, encodeIsDraining, isDraining)
+			require.NoError(t, err)
+		}
+
+		const firstId = base.SQLInstanceID(1000)
+		createInstance(firstId, true, true)
+		const secondId = base.SQLInstanceID(1001)
+		createInstance(secondId, true, false)
+		const thirdId = base.SQLInstanceID(1002)
+		createInstance(thirdId, false, false)
+
+		// Here we show old readers (SQL, instancereader) doesn't break if
+		// they find new schema value before cluster upgrade. This gives extra
+		// confidence on top of version guard we already have.
```

Comment:
nit: doesn't break -> don't break

One note is that here by "old readers" we mean readers on the old _cluster version_. But it doesn't test readers on the old binary version.